### PR TITLE
Fix undefined index at last page of memberlist

### LINKED ISSUE #6072: The database value you're trying to insert does not exist: real_name_high
#### Description
Error message: 
The database value you're trying to insert does not exist: real_name_high

### Steps to reproduce
1. Go to Members | View All Members
2. Click on the last page in a long list - specifically the one after the ellipsis:
image

3. You get the error:
image

### Environment (complete as necessary)
- Version/Git revision: Current
- Database Type: MySQL
- Database Version: 5.7
- PHP Version: 7.3

### Additional information/references
I believe this is an old 2.0 issue as well.  This is the first time I can consistently reproduce it.

### PR DESCRIPTION
**Pull request by sbulen**
Fixes #6072

### CODE REVIEW

**Comment by sbulen** [comment]
Just an FYI - this same code & issue exists in 2.0 (it's reported regularly in my forum's logs...).  I don't think it's worth fixing in 2.0, though.

**Comment by sbulen** [comment]
This bug has been oddly interesting & tricky to reproduce.  It's a function of the # of users you have and your defaultMaxMembers (members to display per page).  That second_offset calculation sometimes goes one cachesize (hardcoded to 500) too high, due to the ceiling function in the calculation.  This is easy to reproduce if you have very high defaultMaxMembers, say 200 or 300.  But, depending on the # of active users total, it can happen even at much lower values of defaultMaxMembers, _if you end up with only ONE member on the very last page..._

Bottom line, the solution is to set the max when the cache is built & make sure you never exceed it.

File: Sources/Memberlist.php [hunk 1]
```diff
@@ -304,7 +304,11 @@ function MLAll()
 	if ($use_cache && $_REQUEST['sort'] === 'real_name' && !isset($_REQUEST['desc']))
 	{
 		$first_offset = $_REQUEST['start'] - ($_REQUEST['start'] % $cache_step_size);
+		if ($first_offset < 0)
+			$first_offset = 0;
```
Comment (Sesquipedalian) [inline Sources/Memberlist.php:308]:
```suggestion
		$first_offset = max(0, $_REQUEST['start'] - ($_REQUEST['start'] % $cache_step_size));
```

File: Sources/Memberlist.php [hunk 2]
```diff
@@ -304,7 +304,11 @@ function MLAll()
...
 		$second_offset = ceil(($_REQUEST['start'] + $modSettings['defaultMaxMembers']) / $cache_step_size) * $cache_step_size;
+		if ($second_offset >= $memberlist_cache['num_members'])
+			$second_offset = $memberlist_cache['num_members'] - 1;
```
Comment (Sesquipedalian) [inline Sources/Memberlist.php:311]:
```suggestion
		$second_offset = min($memberlist_cache['num_members'] - 1, ceil(($_REQUEST['start'] + $modSettings['defaultMaxMembers']) / $cache_step_size) * $cache_step_size);
```

File: Sources/Memberlist.php [hunk 3]
```diff
@@ -319,6 +323,8 @@ function MLAll()
 		if ($first_offset < 0)
 			$first_offset = 0;
 		$second_offset = ceil(($memberlist_cache['num_members'] - $_REQUEST['start']) / $cache_step_size) * $cache_step_size;
+		if ($second_offset >= $memberlist_cache['num_members'])
+			$second_offset = $memberlist_cache['num_members'] - 1;
```
Comment (Sesquipedalian) [inline Sources/Memberlist.php:327]:
```suggestion
		$second_offset = min($memberlist_cache['num_members'] - 1, ceil(($memberlist_cache['num_members'] - $_REQUEST['start']) / $cache_step_size) * $cache_step_size);
```

**Review by Sesquipedalian — CHANGES_REQUESTED** [review]
Using min() and max() will make this cleaner. 🙂

File: Sources/Memberlist.php (hunk 1 above)
Comment (sbulen) [reply Sources/Memberlist.php:308]: Cleaner?  I disagree.  It is much easier to read what it's doing with the comparison to 0...

File: Sources/Memberlist.php (hunk 2 above)
Comment (sbulen) [reply Sources/Memberlist.php:311]: Again, I disagree...  It is not at all clear to see what you're aiming at here...  The suggested is very simple...   

Syntax-wise, 6 of one, half dozen of the other.  Readability-wise, I think we should not make the change.

File: Sources/Memberlist.php (hunk 3 above)
Comment (sbulen) [reply Sources/Memberlist.php:327]: Same as above.

File: Sources/Memberlist.php (hunk 1 above)
Comment (Sesquipedalian) [reply Sources/Memberlist.php:308]: _De gustibis non est disputandum_, I suppose. I personally find min() and max() to be clear and concise, and I prefer to avoid unnecessary `if` statements when I can. But if you have strong feelings about this, I won't dig my heels in. :)

Comment (sbulen) [reply Sources/Memberlist.php:308]: I think it's "gustibus"?

Not strong feelings, just don't like unnecessary changes.  Especially the requisite retests.  (And I honestly think it's less readable...  😉)

I admit my OCD side wants to redo it...  Although these range checks work, in a way, they just mask the real issues with the calc.  I think there are multiple problems: I suspect an off-by-one issue which is why the ceiling sometimes goes too high, DESC sorts have a different count of names on the last (now first...) page, & it fails pretty bad with large page sizes.   

But...  It works.   And the new range checks prevent it from erroring out no matter what.  And there are very few reports of these errors on the support forum anyway.  

I.e., it's a nit...
